sklearn/preprocessing/data.py

```diff
@@ -54,41 +58,82 @@
 )
 
 
-def _mean_and_std(X, axis=0, with_mean=True, with_std=True):
-    """Compute mean and std deviation for centering, scaling.
+def _incremental_mean_and_var(new_X, axis=0, last_mean=.0, last_var=None,
+                              last_n_samples_seen=0):
+    """Compute incremental mean and var for centering, scaling.
 
-    Zero valued std components are reset to 1.0 to avoid NaNs when scaling.
+    last_mean and last_var are the statistics computed at the last step by this
+    function. Both must be initilized to 0.0. In case no scaling is required
+    (caller has with_std=False), last_var can be None. The mean is always
+    required and returned (even if caller has with_mean=True) because it is
+    necessary for the calculation of the var. last_n_samples_seen is the number
+    of samples encountered until now.
     """
-    X = np.asarray(X)
-    Xr = np.rollaxis(X, axis)
+    new_Xr = np.rollaxis(new_X, axis)
 
-    if with_mean:
-        mean_ = Xr.mean(axis=0)
+    if last_var is not None:
+        updated_mean, updated_var, updated_n = \
+            _batch_mean_variance_update(new_Xr, last_mean=last_mean,
+                                        last_variance=last_var,
+                                        last_sample_count=last_n_samples_seen)
     else:
-        mean_ = None
+        new_n = new_Xr.shape[0]
+        updated_n = last_n_samples_seen + new_n
+        updated_mean = last_mean * last_n_samples_seen + new_Xr.sum(axis=0)
+        updated_mean /= updated_n
+        updated_var = None
 
-    if with_std:
-        std_ = Xr.std(axis=0)
-        std_ = _handle_zeros_in_scale(std_)
+    return updated_mean, updated_var, updated_n
+
+
+def _incremental_min_and_max(new_X, axis=0, last_min=None, last_max=None):
+    """Compute incremental min and max.
+
+    last_min, last_max are the statistics computed until now.
+    Both are initilized as None.
+    """
+
+    if last_min is not None:
+        updated_min = np.minimum(last_min, np.min(new_X, axis))
+    else:
+        updated_min = np.min(new_X, axis)
+
+    if last_max is not None:
```

Comment:
Since you have added this, I think you can remove the hasattr(self, "n_samples_") check in `MinMaxScaler` right? (or vice versa, keep that and remove this check.

But to be consistent with `AbsMaxScaler` I think it might be better to remove the None condition.